File: python/0133_clone_graph.py
from collections import deque


# Definition for a Node.
class Node:
    def __init__(self, val=0, neighbors=None):
        if neighbors is None:
            neighbors = []
        self.val = val
        self.neighbors = neighbors


class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None
        queue = deque([node])
        res_root = Node(node.val)
        # res_queue = deque([res_root])
        traversed = {node.val: res_root}
        while queue:
            cur = queue.popleft()
            # res_cur = res_queue.popleft()
            for neighbor in cur.neighbors:
                if neighbor.val in traversed.keys():
                    res_neighbor = traversed[neighbor.val]
                else:
                    res_neighbor = Node(neighbor.val)
                    traversed[neighbor.val] = res_neighbor
                    queue.append(neighbor)
                    # res_queue.append(res_neighbor)
                traversed[cur.val].neighbors.append(res_neighbor)
        return res_root


a = Node(0)
b = Node(1)
c = Node(2)
a.neighbors.append(b)
a.neighbors.append(c)
b.neighbors.append(a)
b.neighbors.append(c)
c.neighbors.append(a)
c.neighbors.append(b)

o = Solution().cloneGraph(a)
las = 1

File: cpp/0133_clone_graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
  // unordered_map<Node*, Node*>visited;
    Node* visited[101]{nullptr};
    // unordered_map<int, Node*>visited;
    Node* cloneGraph(Node* node) {
      if (node == nullptr) 
        return node;
      // cout << "v: " << visited[node] << endl;
      if(visited[node->val] != nullptr) {
        return visited[node->val];
      }

      Node* cloned_node = new Node(node->val);
      visited[node->val] = cloned_node;
      for (int i = 0 ; i < node->neighbors.size(); i++) {
        cloned_node->neighbors.emplace_back(cloneGraph(node->neighbors[i]));
      }
      return cloned_node;
    }
};
